content/6_Advanced/Wavelet.mdx

```diff
@@ -32,9 +30,342 @@ Like a segment tree on values rather than indices.
+  child
+- If $\texttt{left\_occurrences}$ is less than $k$, it means the
+  $k$-th smallest element is in the right subtree. We adjust k by subtracting
+  $\texttt{left\_occurrences}$ from $k$, update our range, and recurse into the right child
+
+<Info title="Notice">
+  Notice we still update $l, r$ accordingly when we go left or right
+</Info>
+
+The answer then will be the value of the node we end up on (leaf).
+
+In conclusion we maintain our ranges l and r as we recurse down to our child, and when we reach the child node we can return $r$ - $l$.
+
+## Implementation
+**Time Complexity:** $\mathcal{O}(Q \cdot \log(M))$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+
+using namespace std;
+constexpr int MAX_VAL = 1e9 + 2;
+
+struct Segment {
+	Segment *left = nullptr, *right = nullptr;
+	int l, r, mid;
+	bool children = false;
+	vector<pair<int, int>> indices;  // Index, Value
+	vector<int> prefix_b;
+
+	Segment(int l, int r, const vector<pair<int, int>> &indices)
+	    : l(l), r(r), mid((r + l) / 2), indices(indices) {
+
+		// Calculate the prefix B
+		int i = 1;
+		int j = 0;
+		prefix_b.resize(indices.size() + 1);
+		for (auto [ind, val] : indices) {
+			if (val < mid) j++;
+			prefix_b[i++] = j;
+		}
+	}
+
+	// Sparse since values can go up to 1e9
+	void update() {
+		if (children) { return; }
+		children = true;
+		if (r - l > 1) {
+			// Split the indices for left and right child
+			vector<pair<int, int>> leftIndices, rightIndices;
+			partition_copy(indices.begin(), indices.end(), leftIndices.begin(),
+			               rightIndices.begin(), [this](const pair<int, int> &elem) {
+				               return elem.second < mid;
+			               });
+
+			left = new Segment(l, mid, leftIndices);
+			right = new Segment(mid, r, rightIndices);
+		}
+	}
+
+	int find_k_smallest(int a, int b, int k) {
+		update();
+		if (r - l <= 1) { return l; }
+
+		int lb = prefix_b[a];
+		int lr = prefix_b[b];
+		int inLeft = lr - lb;  // Amount of values in range (a,b) that are less the mid
+
+		if (k <= inLeft) {
+			return left->find_k_smallest(lb, lr, k);  // Appears in left
+		} else {
+			return right->find_k_smallest(a - lb, b - lr,
+			                              k - inLeft);  // Appears in right
+		}
+	}
+};
+
+int main() {
+	int n, q;
+	cin >> n >> q;
+
+	vector<pair<int, int>> indices;
+	for (int i = 0; i < n; ++i) {
+		int v;
+		cin >> v;
+		indices.emplace_back(i, v);
+	}
+	Segment seg(0, MAX_VAL, indices);
+
+	for (int i = 0; i < q; ++i) {
+		int a, b, k;
+		cin >> a >> b >> k;
+		k++;
+		cout << seg.find_k_smallest(a, b, k) << " ";
+	}
+}
+```
+</CPPSection>
+</LanguageSection>
+
+
+
+## Supporting updates
+
+Let's support updates of type:
+	- change value at index $i$ to $y$
+
+We can traverse down to the leaf to remove the old element and also traverse down to add the new element.
+
+So what do the updates change?
+ -
+ Our indices vector
+ Our prefix vector
+
```

Comment:
are you previewing this in usaco.guide/editor or another markdown editor?

if you're attempting a question->answer section, use <Spoiler>

Reply:
Use what?